designs/forLoopIteratorsDesign.md

```diff
@@ -0,0 +1,270 @@
+## Test Cases ##
+* **Test Case 1**: for loop with range and one parameter in range
+ 
+Input:
+```
+i : int = 0
+for i in range(10):
+    print(i)
+```
+Output:
+```
+0
+1
+2
+3
+4
+5
+6
+7
+8
+9
+```
+* **Test Case 2**: for loop with range and two parameters in range
+ 
+Input:
+```
+i : int = 0
+for i in range(3,6):
+    print(i)
+```
+Output:
+```
+3
+4
+5
+```
+ 
+* **Test Case 3**: for loop with range and three parameters in range
+ 
+Input:
+```
+i : int = 0
+for i in range(0,10,2):
+    print(i)
+```
+Output:
+```
+0
+2
+4
+6
+8
+```
+* **Test Case 4**: for loop with break in the main for loop body
+ 
+Input:
+```
+i : int = 0
+for i in range(10):
+    print(i)
+    break
+```
+Output:
+```
+0
+```
+* **Test Case 5**:  for loop with break inside a if body
+ 
+Input:
+```
+i : int = 0
+for i in range(10):
+    If i\>5:
+        break
+    print(i)
+```
+Output:
+```
+0
+1
+2
+3
+4
+5
+```
+ 
+* **Test Case 6**:  for loop with continue inside the main for body
+ 
+Input:
+```
+i : int = 0
+for i in range(5):
+    print(i*100)
+    continue
+    print(i)
+```
+Output:
+```
+0
+100
+200
+300
+400
+```
+ 
+* **Test Case 7**:  for loop with continue inside a if body
+ 
+Input:
+```
+i : int = 0
+for i in range(10):
+    If i%2:
+        continue
+    print(i)
+```
+Output:
+```
+0
+2
+4
+6
+8
+```
+ 
+* **Test Case 8**: for else construct
+ 
+Input:
+```
+i : int = 0
+for i in range(10):
+    If i\>5:
+        break
+    print(i)
+else:
+    print(“i was greater than 5”) // will only work if basic strings are implemented, we can change this to print(i*100) or something
+```
+Output:
+```
+0
+1
+2
+3
+4
+5
+i was greater than 5
+```
+* **Test Case 9**: operations on range
+ 
+Input:
+```
+r : range = None
+r = range(0, 20, 2)
+print(11 in r)
+```
+Output:
+```
+False
+```
+* **Test Case 10**: inbuilt functions on range, index() returning Value Error
+ 
+Input:
+```
+r : range = None
+r = range(0, 20, 2)
+print(r.index(11))
+```
+Output:
+```
+ValueError: 11 is not in range
+```
+* **Test Case 11**:  inbuilt functions on range, index() returning index of reachable number
+ 
+Input:
+```
+r : range = None
+r = range(0, 20, 2)
+print(r.index(10))
+```
+Output:
+```
+5
+```
+* **Test Case 12**: type checking for range parameters
+ 
+Input:
+```
+range(0, 10.2, 0.2)
+```
+Output:
+```
+TypeError: float object cannot be interpreted as integer
+```
+* **Test Case 13**: type checking for range parameters
+ 
+Input:
+```
+range(0, 10, 1, 2)
+```
+Output:
+```
+TypeError: range expected at most 3 arguments, got 4
+```
+* **Test Case 14**: type checking for range parameters
+ 
+Input:
+```
+range()
+```
+Output:
+```
+TypeError: range expected 1 argument, got 0
+```
+* **Test Case 15**: type checking for range parameters
+ 
+Input:
+```
+range()
+```
+Output:
+```
+TypeError: range expected 1 argument, got 0
+```
+## Changes required in AST, IR and builtin libraries ##
+ 
+* The following changes will be required in ast.ts:
+ 
+    * In type Stmt\<A\>:
```

Comment:
How will these be type-checked?

Reply:
We have described it here: https://github.com/tapan0007/chocopy-wasm-compiler-B/blob/2022/designs/forLoopIteratorsDesign.md#changes-required-in-ast-ir-and-builtin-libraries